### PR TITLE
Enable Multiroot for Report Viewer

### PR DESCRIPTION
**Pull request by nestabentum**
This is still WIP but I wanted to share my approach already to receive feedback

This PR intends to fix two major issues (both noted in issue #354):

1. Submission names were concatenated to produce the comparison file name. When JPlag is run with multiple roots, the submission names contain backward slashes causing file creation to crash.
2. As Submission names were concatenated  duplicate names could easily be produced in the single source case ({`A, A-B, C, B-C}` would result in two `A-B-C` )

My approach to fixing this:

 * Differentiation between the submission name, submission id, and submission display name.
      * `Submission name` simply refers to Submission#name
      * `Submission id` is newly introduced and used only in the context of the report viewer creation. Generation of this id is the topic of another (our next 😄 ) discussion, but currently, it is the sanitized version of the submission name (File separator replaced by underscore). This submission id serves as an id for internal handling in the report viewer. 
      * `Display name` is the name of a submission to be displayed by the report viewer. The `overview.json` file contains a map <`submission id` -> `display name`> the report viewer will use to resolve the name for presentation purposes.
 * Two be able to abolish name concatenation of submission names for the comparison file name, I introduced a nested lookup table to the `overview.json`. It has the structure <`submission id `-> <<`submission id`> -><`comparison file name`>>>. With this data structure, there will be no duplicate comparison file names and additionally, the comparison file name can be chosen arbitrarily as the frontend will only have to execute two gets two receive the file name and will not have to rely on string concatenation.

Currently, the nested lookup table contains all submission ids in its first level, and in the second level, it contains each comparison for the corresponding submission id. This redundancy could be eliminated and actually halve the table's size but this is an optimization for later.
The UI isn't fully compatible with this approach yet, but will be with my next commits.

EDIT: 
The report viewer is now fully compatible with the updated file format.
I realised that the problem in 2. persisted, as I still used id concatenation. Therefore, I changed the comparison file name generation to simply be a number that is counted up for each file. While this worsens human readability, it solves the aforementioned issue. As always, suggestions on how to solve this issue while maintaining readability are welcome.

### CODE REVIEW

File: jplag/src/main/java/de/jplag/reporting/jsonfactory/DirectoryCreator.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    /**
+     * Creates a directory.
+     * @param path The path under which the new directory ought to be created
+     * @param name The name of the new directory
+     * @return The created directory
+     */
+    public static File createDirectory(String path, String name) {
+        File directory = new File(path.concat("/").concat(name));
+        if (!directory.exists() && !directory.mkdirs()) {
+            logger.error("Failed to create dir.");
```
Comment (dfuchss) [inline jplag/src/main/java/de/jplag/reporting/jsonfactory/DirectoryCreator.java:20]: If recovery is not possible. You should use exceptions

File: jplag/src/test/java/de/jplag/reporting/reportobject/mapper/ComparisonReportWriterTest.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import de.jplag.JPlagResult;
+import de.jplag.Submission;
+import de.jplag.TestBase;
+import de.jplag.exceptions.ExitException;
+import de.jplag.reporting.jsonfactory.ComparisonReportWriter;
+
+public class ComparisonReportWriterTest extends TestBase {
+
+    @Test
+    public void test() throws ExitException {
```
Comment (dfuchss) [inline jplag/src/test/java/de/jplag/reporting/reportobject/mapper/ComparisonReportWriterTest.java:17]: Maybe use a better name for the test

**Review by dfuchss — COMMENTED** [review]
Very minor comments :)

**Review by sebinside — APPROVED** [review]
I only had a quick look over the code, so there is a more in-depth review from @jplag/studdev required. But it tested the PR with two directories with 200 and 50 submissions, and everything worked fine.

Probably not related to this PR, but sometimes, the comparison viewer still breaks, but not predictable with the warning `Could not find comparison file. indes.js:43` and then with the error:
```
templateLoader.js?e847:50 
        
       Uncaught (in promise) TypeError: _ctx.isAnonymous is not a function
    at Proxy.render (templateLoader.js?e847:50:1)
    at renderComponentRoot (runtime-core.esm-bundler.js?d2dd:896:1)
    at ReactiveEffect.componentUpdateFn [as fn] (runtime-core.esm-bundler.js?d2dd:5580:1)
    at ReactiveEffect.run (reactivity.esm-bundler.js?89dc:185:1)
    at instance.update (runtime-core.esm-bundler.js?d2dd:5694:1)
    at setupRenderEffect (runtime-core.esm-bundler.js?d2dd:5708:1)
    at mountComponent (runtime-core.esm-bundler.js?d2dd:5490:1)
    at processComponent (runtime-core.esm-bundler.js?d2dd:5448:1)
    at patch (runtime-core.esm-bundler.js?d2dd:5038:1)
    at ReactiveEffect.componentUpdateFn [as fn] (runtime-core.esm-bundler.js?d2dd:5660:1)
```

File: jplag/src/main/java/de/jplag/reporting/jsonfactory/DirectoryCreator.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
+package de.jplag.reporting.jsonfactory;
+
+import java.io.File;
+import java.io.IOException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DirectoryCreator {
```
Comment (tsaglam) [inline jplag/src/main/java/de/jplag/reporting/jsonfactory/DirectoryCreator.java:9]: JDoc

File: jplag/src/main/java/de/jplag/reporting/reportobject/ReportObjectFactory.java [hunk 4]
```diff
@@ -19,85 +19,64 @@
...
+        submissionToIdFunction = (Submission submission) -> submissionNameToIdMap.get(submission.getName());
     }
 
-    private static void copySubmissionFilesToReport(String path, JPlagResult result) {
+    private void copySubmissionFilesToReport(String path, JPlagResult result) {
         List<JPlagComparison> comparisons = result.getComparisons(result.getOptions().getMaximumNumberOfComparisons());
         var submissions = getSubmissions(comparisons);
-        var submissionsPath = createDirectory(path, SUBMISSIONS_FOLDER);
+        var submissionsPath = createSubmissionsDirectory(path);
+        if (submissionsPath == null)
```
Comment (tsaglam) [inline jplag/src/main/java/de/jplag/reporting/reportobject/ReportObjectFactory.java:73]: Avoid if statements without curly braces.

File: jplag/src/main/java/de/jplag/reporting/reportobject/ReportObjectFactory.java [hunk 5]
```diff
@@ -109,52 +88,78 @@ private static void copySubmissionFilesToReport(String path, JPlagResult result)
...
-        return submissions;
+    private File createSubmissionDirectory(String path, File submissionsPath, Submission submission) {
+        File directory;
+        try {
+            directory = createDirectory(submissionsPath.getPath(), submissionToIdFunction.apply(submission));
+        } catch (IOException e) {
+            logger.error("Could not create directory " + path + " for report viewer generation", e);
+            return null;
+        }
+        return directory;
```
Comment (tsaglam) [inline jplag/src/main/java/de/jplag/reporting/reportobject/ReportObjectFactory.java:99]:
```suggestion
        try {
            return createDirectory(submissionsPath.getPath(), submissionToIdFunction.apply(submission));
        } catch (IOException e) {
            logger.error("Could not create directory " + path + " for report viewer generation", e);
            return null;
        }
```

File: jplag/src/main/java/de/jplag/reporting/reportobject/ReportObjectFactory.java [hunk 6]
```diff
@@ -109,52 +88,78 @@ private static void copySubmissionFilesToReport(String path, JPlagResult result)
...
-        return result.getComparisons(numberOfComparisons);
+    private File createSubmissionsDirectory(String path) {
+        File submissionsPath;
+        try {
+            submissionsPath = createDirectory(path, SUBMISSIONS_FOLDER);
+        } catch (IOException e) {
+            logger.error("Could not create directory " + path + " for report viewer generation", e);
+            return null;
+        }
+        return submissionsPath;
```
Comment (tsaglam) [inline jplag/src/main/java/de/jplag/reporting/reportobject/ReportObjectFactory.java:110]: As above.

File: jplag/src/main/java/de/jplag/reporting/reportobject/mapper/SubmissionNameToIdMapper.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+import java.io.File;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import de.jplag.JPlagComparison;
+import de.jplag.JPlagResult;
+import de.jplag.Submission;
+
+public class SubmissionNameToIdMapper {
```
Comment (tsaglam) [inline jplag/src/main/java/de/jplag/reporting/reportobject/mapper/SubmissionNameToIdMapper.java:12]: JDoc

File: jplag/src/main/java/de/jplag/reporting/reportobject/mapper/SubmissionNameToIdMapper.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+        HashMap<String, String> idToName = new HashMap<>();
+        getComparisons(result).forEach(comparison -> {
+            idToName.put(comparison.getFirstSubmission().getName(), sanitizeNameOf(comparison.getFirstSubmission()));
+            idToName.put(comparison.getSecondSubmission().getName(), sanitizeNameOf(comparison.getSecondSubmission()));
+        });
+        return idToName;
+    }
+
+    private static String sanitizeNameOf(Submission comparison) {
+        return comparison.getName().replace(File.separator, "_");
```
Comment (tsaglam) [inline jplag/src/main/java/de/jplag/reporting/reportobject/mapper/SubmissionNameToIdMapper.java:28]: Ideally `"_"` should be a constant.

**Review by tsaglam — CHANGES_REQUESTED** [review]
Just a few minor things...

**Review by tsaglam — APPROVED** [review]
Looks good! @nestabentum ready to merge from your end?

**Comment by nestabentum** [comment]
@tsaglam this is good to go

**Comment by robinmaisch** [comment]
Sorry, I'm late to the party, but I noticed that the `ComparisonReportWriterTest` writes test report files to `${baseDir}\jPlag` and does not remove them afterwards. I suggest that they are created in a special output directory and/or removed after the test.

**Comment by tsaglam** [comment]
> 

Should be already fixed by #607

**Pull request merged**